File: stasys.py
#! /usr/bin/env python3

###############################################################################
# Author: Stephen P Marsden
# Date: 2022-09-06
#
# Description: Outputs a single line summary of key CPU, RAM, and Swap details, 
#              intended to be embedded in the GNOME topbar.
#
# Dependencies: python3
#               numpy
#               ip
#               sensors
#               vmstat
#               free
#
###############################################################################

# For gettinf the current unix time.
from time import time
# For executing shell commands.
from subprocess import Popen, PIPE, DEVNULL
# for quicker array calculations.
from numpy import array

def t_rx_tx_Mb():
    """
    Retreives the current recorded received and transmitted bits, along with the
    unix time at which there were accessed. Subsequent calls can be used to 
    calculate the rate of data transfer.
    
    Args:
        None
    
    Returns:
        float - unix timestamp [s]
        int   - received bits [Mb]
        int   - transmitted bits [Mb]
    
    Raises:
        None
    
    ToDo:
        None
    """

    # Get the unix timestamp.
    unix_time = time()

    # Run ip to get the received / transmitted bits.
    process = Popen(['ip', '-s','link'], 
                           stdout=PIPE, 
                           stderr=DEVNULL,
                           universal_newlines=True)
    # Read the output.
    stdout = [line for line in process.stdout.readlines()]

    # Remove the loopback interface.
    # Fins the start of the loopback.
    skip_start = None
    for i,line in enumerate(stdout):
        if line[0] != " ":
            if "LOOPBACK" in line:
                skip_start = i
                break
    # Find the end of the loopback.
    skip_end = None
    for i,line in enumerate(stdout[skip_start+1:]):
        if line[0] != " ":
            skip_end = skip_start + i
            break
    # Remove the loopback entry.
    stdout = stdout[0:skip_start] + stdout[skip_end+1:]
    
    # Split each line by whitespace.
    stdout = [line.split() for line in stdout]

    # Get the received/transmitted line numbers.
    rx_lines = [i+1 for i in range(len(stdout)) if stdout[i][0]=="RX:"]
    tx_lines = [i+1 for i in range(len(stdout)) if stdout[i][0]=="TX:"]

    # Sum the received/transmitted bits and convert to Mb.
    rx_sum_Mb = sum([int(stdout[i][0]) for i in rx_lines]) * 8 / 1e6
    tx_sum_Mb = sum([int(stdout[i][0]) for i in tx_lines]) * 8 / 1e6

    return array([unix_time, rx_sum_Mb, tx_sum_Mb])

def cpu_freq_MHz():
    """
    Reads in the current CPU frequencies and calculates the average, returning
    a single frequency.
    
    Args:
        None
    
    Returns:
        float - Frequency in MHz.
    
    Raises:
        None
    
    ToDo:
        None
    """

    # Storage for all the CPU frequencies.
    frequencies_MHz = []
    # Read in the file containing the values.
    with open("/proc/cpuinfo","r") as cpuinfo:
        # Find each line containing a CPU frequency.
        for line in cpuinfo.readlines():
            if line.startswith("cpu MHz"):
                # Record the frequency.
                frequencies_MHz.append(float(line.split()[-1]))

    # Get the average frequency.
    return sum(frequencies_MHz) / len(frequencies_MHz)

def main():

    ## General #################################################################

    # Get the sensors output. This will be used for temperatures and fan speed.
    sensors_process = Popen(['sensors'],
                            stdout=PIPE,
                            stderr=DEVNULL,
                            universal_newlines=True)
    # Read in the output from sensors.
    sensors_stdout = [line.split() for line in sensors_process.stdout.readlines()]
    

    ## Network Usage 1/2 #######################################################

    # Record the time and cumulative network usage.
    #   0: unix timestamp.
    #   1: RX in Mb
    #   2: TX in Mb
    t_rx_tx_start = t_rx_tx_Mb()


    ## CPU #####################################################################

    ### CPU Usage.
    # Run vmstat to get the idle percentage.
    vmstat_process = Popen(['vmstat', '1','1'], 
                           stdout=PIPE, 
                           stderr=DEVNULL,
                           universal_newlines=True)
    # Read the output.
    vmstat_stdout = [line.split() for line in vmstat_process.stdout.readlines()]
    # Find the index of the idle percentage.
    id_index = vmstat_stdout[1].index("id")
    # Get the inverse of the idle percentage.
    cpu_percentage = 100 - int(vmstat_stdout[2][id_index])

    ### CPU Frequency.
    # Get the average CPU frequency.
    freq_MHz = cpu_freq_MHz()
    # Format the output string.
    cpu_freq_str = ""
    if freq_MHz > 1e3:
        cpu_freq_str = f"{freq_MHz/1e3:.1f} GHz"
    else:
        cpu_freq_str = f"{round(freq_MHz)} MHz"

    ### CPU Temperature.
    cpu_temps = [line[2] for line in sensors_stdout if len(line)>0 and line[0]=="Core"]
    # Take the average temperature.
    cpu_temps_average = round(sum([float(value[1:-2]) for value in cpu_temps]) / len(cpu_temps))
    
    
    ## Memory Usage ############################################################

    ### Gather data.
    # Run free to gather memory information.
    free_process = Popen(['free', '-m'], 
                           stdout=PIPE, 
                           stderr=DEVNULL,
                           universal_newlines=True)
    # Read the output.
    free_stdout = [line.split() for line in free_process.stdout.readlines()]
    # Add an empty entry at the start of the title row to keep everything in line.
    free_stdout[0] = [''] + free_stdout[0]

    ### Get the column indexes.
    # Total capacity.
    total_index = free_stdout[0].index("total")
    # Capacity used.
    used_index = free_stdout[0].index("used")

    ### RAM
    # Find the RAM index.
    ram_index = [line[0] for line in free_stdout].index("Mem:")
    # Get the RAM values.
    ram_total = int(free_stdout[ram_index][total_index])
    ram_used  = int(free_stdout[ram_index][used_index ])
    # Get the RAM used percentage.
    ram_percentage = round(100 * ram_used/ram_total)

    ### Swap
    swap_index = [line[0] for line in free_stdout].index("Swap:")
    if swap_index < len(free_stdout):
        # Get the swap values.
        swap_total = int(free_stdout[swap_index][total_index])
        swap_used  = int(free_stdout[swap_index][used_index ])
        # Get the RAM used percentage.
        if swap_total > 0:
            swap_percentage = round(100 * swap_used/swap_total)

    ## Network Usage 2/2 #######################################################

    # Record the time and cumulative network usage.
    #   0: unix timestamp.
    #   1: RX in Mb
    #   2: TX in Mb
    t_rx_tx_stop = t_rx_tx_Mb()

    # Calculate the transfer rate.
    delta_t_rx_tx = t_rx_tx_stop - t_rx_tx_start
    drx_dt = f"{delta_t_rx_tx[1] / delta_t_rx_tx[0]:.1f}"
    dtx_dt = f"{delta_t_rx_tx[2] / delta_t_rx_tx[0]:.1f}"

    ## Output ##################################################################

    # Define the separator.
    separator = "   |   "

    # Put the output together.
    output = ""
    output += f"CPU:   {cpu_freq_str}   {cpu_percentage}%   {cpu_temps_average}°C"
    output += separator
    output += f"RAM: {ram_percentage}%"
    if swap_total > 0:
        output += separator
        output += f"Swap: {swap_percentage}%"
    output += separator
    output += f"↑{dtx_dt} Mb/s   ↓{drx_dt} Mb/s"


    print(output)
    return

if __name__=="__main__":
    main()

File: stasys.go
package main

import (
	"bytes"
	"fmt"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"time"
)

func stdout2fields(stdout bytes.Buffer) [][]string {
	// Convert a byte stream from a command line programme into a 2D slice
	// split first by \n, then by whitespace.

	// Split by \n
	stdout_lines := strings.Split(stdout.String(), "\n")

	// Split into fields.
	var fields [][]string
	for _, line := range stdout_lines {
		fields = append(fields, strings.Fields(line))
	}

	return fields
}

func t_rx_tx_Mb() (float64, float64, float64) {

	// Record the time that the data was read in.
	read_time := float64(time.Now().UnixNano()) / 1e9

	// Run ip to get the received / transmitted bits.
	// Define the command.
	cmd := exec.Command("ip", "-s", "link")
	// Add a stream for the stdout.
	var ip_stdout bytes.Buffer
	cmd.Stdout = &ip_stdout
	// Run the command.
	_ = cmd.Run()
	// 4 ms.

	// Split the output into lines.
	ip_lines := strings.Split(ip_stdout.String(), "\n")
	// Storage for the current interface name.
	interface_name := ""
	// Storage for if we want to do something with the line that follows.
	is_rx := false
	is_tx := false
	// Storage for the cumulative RX and TX vaues.
	rx := 0
	tx := 0
	for _, line := range ip_lines {
		// Split each line into fields.
		ip_fields := strings.Fields(line)
		// Check if the line is an interface heading.
		if len(ip_fields) > 2 && ip_fields[2][0:1] == "<" {
			// Strip formatting away from interface name.
			interface_name = strings.Split(ip_fields[2], ",")[0][1:]
		} else if interface_name == "LOOPBACK" {
			// Skip the LOOPBACK interface.
			continue
		} else if is_rx {
			// Read the RX bits for this interface.
			drx, _ := strconv.Atoi(ip_fields[0])
			// Add the bits to the running total.
			rx += drx
			// Reset the flag ready for the next line.
			is_rx = false
		} else if is_tx {
			// Read the TX bits for this interface.
			dtx, _ := strconv.Atoi(ip_fields[0])
			// Add the bits to the running total.
			tx += dtx
			// Reset the flag ready for the next line.
			is_tx = false
		} else if len(ip_fields) > 0 {
			// Check if the next line is RX.
			if ip_fields[0] == "RX:" {
				is_rx = true
			}
			// Check if the next line is TX.
			if ip_fields[0] == "TX:" {
				is_tx = true
			}
		}
	}

	// Convert from B to Mb.
	rx_Mb := float64(rx) * 8 / 1e6
	tx_Mb := float64(tx) * 8 / 1e6

	return read_time, rx_Mb, tx_Mb
}

func cpu_freq_MHz() float64 {
	// Storage for the average frequency.
	average_frequency := 0.
	// Record how many CPUs were measured.
	n_cpu := 0.
	// Read in the file containing the frequencies.
	data, _ := os.ReadFile("/proc/cpuinfo")
	// Record the CPU frequencies.
	for _, line := range strings.Split(string(data), "\n") {
		// Split the line by whitespace.
		fields := strings.Fields(line)
		// If the line contains CPU information, record it.
		if len(fields) > 3 && fields[0] == "cpu" && fields[1] == "MHz" {
			// Add the frequency to the total.
			dfreq, _ := strconv.ParseFloat(fields[3], 64)
			average_frequency += dfreq
			// Record that another CPU has been read.
			n_cpu += 1
		}
	}
	// Calculate the average frequency.
	average_frequency /= n_cpu
	// Return that average.
	return average_frequency
}

func main() {

	// General /////////////////////////////////////////////////////////////////

	// Get the sensors output. This will be used for temperatures and fan speed.
	// Define the command.
	cmd_sensors := exec.Command("sensors")
	// Add a stream for the stdout.
	var sensors_stdout bytes.Buffer
	cmd_sensors.Stdout = &sensors_stdout
	// Run the command.
	_ = cmd_sensors.Run() // 14 ms.

	// Split the output.
	sensors := stdout2fields(sensors_stdout)

	// Network Usage 1/2 ///////////////////////////////////////////////////////

	// Record the time and cumulative network usage.
	network_t_start_s, network_rx_start_Mb, network_tx_start_Mb := t_rx_tx_Mb()

	// CPU /////////////////////////////////////////////////////////////////////

	// CPU Usage.
	// Run vmstat to get the idle percentage.
	// Define the command.
	cmd_vmstat := exec.Command("vmstat", "1", "1")
	// Add a stream for the stdout.
	var vmstat_stdout bytes.Buffer
	cmd_vmstat.Stdout = &vmstat_stdout
	// Run the command.
	_ = cmd_vmstat.Run() // 6 ms.

	// Split output into lines.
	vmstat_lines := strings.Split(vmstat_stdout.String(), "\n")
	// Find the index of the idle percentage.
	id_index := -1
	for i, heading := range strings.Fields(vmstat_lines[1]) {
		if heading == "id" {
			id_index = i
			break
		}
	}
	// Get the data divided into fields.
	vmstat_data := strings.Fields(vmstat_lines[2])
	// Get the idle percentage.
	id_percentage, _ := strconv.Atoi(vmstat_data[id_index])
	// Get the complement of the idle percentage.
	cpu_percentage := 100 - id_percentage

	// CPU Frequency.
	// Get the average CPU frequency.
	freq_MHz := cpu_freq_MHz()
	// Format the output string.
	cpu_freq_str := ""
	if freq_MHz > 1e3 {
		cpu_freq_str = fmt.Sprintf("%.1f GHz", freq_MHz/1e3)
	} else {
		cpu_freq_str = fmt.Sprintf("%.0f MHz", freq_MHz)
	}

	// CPU Temperature.
	cpu_temp := 0.
	// Count the number of cores being averaged over.
	n_cpu := 0
	// Get the sum of the core temperatures.
	for _, line := range sensors {
		if len(line) > 0 && line[0] == "Core" {
			temp := line[2][1 : len(line[2])-3]
			dt, _ := strconv.ParseFloat(temp, 64)
			cpu_temp += dt
			n_cpu++
		}
	}
	cpu_temp /= float64(n_cpu)

	// Fan Speed ///////////////////////////////////////////////////////////////
	// for _, line := range sensors {
	// 	if len(line) > 0 && line[0] ==
	// }

	// Memory Usage ////////////////////////////////////////////////////////////

	// Run free to get the memory stats.
	// Define the command.
	cmd_free := exec.Command("free", "-m")
	// Add a stream for the stdout.
	var free_stdout bytes.Buffer
	cmd_free.Stdout = &free_stdout
	// Run the command.
	_ = cmd_free.Run() // 6 ms.

	// Split the output.
	free := stdout2fields(free_stdout)

	// Find the index of the total and free columns.
	used_index := -1
	total_index := -1
	for i, entry := range free[0] {
		if entry == "total" {
			total_index = i + 1
		} else if entry == "used" {
			used_index = i + 1
		}
		if used_index > -1 && total_index > -1 {
			break
		}
	}

	// Find the index of the memory entry.
	mem_index := -1
	swap_index := -1
	for i, line := range free {
		if len(line) == 0 {
			continue
		}
		if line[0] == "Mem:" {
			mem_index = i
		} else if line[0] == "Swap:" {
			swap_index = i
		}
	}

	mem_used, _ := strconv.ParseFloat(free[mem_index][used_index], 64)
	mem_total, _ := strconv.ParseFloat(free[mem_index][total_index], 64)
	mem_percentage := 0.
	if mem_total > 0 {
		mem_percentage = 100. * mem_used / mem_total
	}

	swap_percentage := -1.
	if swap_index > -1 {
		swap_used, _ := strconv.ParseFloat(free[swap_index][used_index], 64)
		swap_total, _ := strconv.ParseFloat(free[swap_index][total_index], 64)
		if swap_total > 0 {
			swap_percentage = 100. * swap_used / swap_total
		}
	}

	// Network Usage 2/2 ///////////////////////////////////////////////////////

	// Record the time and cumulative network usage.
	network_t_stop_s, network_rx_stop_Mb, network_tx_stop_Mb := t_rx_tx_Mb()

	// Calculate the transfer rate.
	network_t_delta_s := network_t_stop_s - network_t_start_s
	network_rx_delta_Mb := network_rx_stop_Mb - network_rx_start_Mb
	network_tx_delta_Mb := network_tx_stop_Mb - network_tx_start_Mb
	drx_dt := network_rx_delta_Mb / network_t_delta_s
	dtx_dt := network_tx_delta_Mb / network_t_delta_s

	// Output //////////////////////////////////////////////////////////////////

	divider := " | "

	output := ""
	output += fmt.Sprintf("CPU: %s %d%% %.0f°C",
		cpu_freq_str,
		cpu_percentage,
		cpu_temp)
	output += divider
	output += fmt.Sprintf("RAM: %.0f%%", mem_percentage)
	if swap_percentage > -1 {
		output += divider
		output += fmt.Sprintf("Swap: %.0f%%", swap_percentage)
	}
	output += divider
	output += fmt.Sprintf("↑%.1f Mb/s ↓%.1f Mb/s", dtx_dt, drx_dt)

	fmt.Println(output)

}
